applications/serial_lte_modem/src/slm_at_fota.c

```diff
@@ -0,0 +1,299 @@
+#include "slm_at_fota.h"
+
+LOG_MODULE_REGISTER(fota, CONFIG_SLM_LOG_LEVEL);
+
+/* fuile_uri: scheme://[user:password@]hostname[:port]path */
+#define FILE_URI_MAX	256
+#define SCHEMA_HTTP	"http"
+#define SCHEMA_HTTPS	"https"
+#define URI_HOST_MAX	64
+#define URI_PATH_MAX	128
+
+#define APN_MAX		32
+
+#define AT_FOTA	"AT#XFOTA"
+
+/* Some features need fota_download update */
+#define FOTA_FUTURE_FEATURE	0
+
+enum slm_fota_operation {
+	AT_FOTA_ERASE,
+	AT_FOTA_START,
+	AT_FOTA_PAUSE_RESUME,
+	AT_FOTA_STOP
+};
+
+/* global functions defined in different files */
+void rsp_send(const uint8_t *str, size_t len);
+
+/* global variable defined in different files */
+extern struct at_param_list at_param_list;
+extern char rsp_buf[CONFIG_AT_CMD_RESPONSE_MAX_LEN];
+extern struct k_work_q slm_work_q;
+
+static int do_fota_erase(void)
+{
+	int fd;
+	int err;
+	nrf_dfu_fw_offset_t offset;
+	nrf_socklen_t len = sizeof(offset);
+
+	LOG_INF("Erasing scratch");
+
+	fd = nrf_socket(NRF_AF_LOCAL, NRF_SOCK_STREAM, NRF_PROTO_DFU);
+	if (fd < 0) {
+		LOG_ERR("nrf_socket error: %d", errno);
+		return -errno;
+	}
+	err = nrf_setsockopt(fd, NRF_SOL_DFU, NRF_SO_DFU_BACKUP_DELETE,
+			NULL, 0);
+	if (err) {
+		LOG_ERR("nrf_setsockopt error: %d", errno);
+		nrf_close(fd);
+		return -errno;
+	}
+	while (true) {
+		err = nrf_getsockopt(fd, NRF_SOL_DFU, NRF_SO_DFU_OFFSET,
+			&offset, &len);
+		if (err < 0) {
+			k_sleep(K_SECONDS(1));
+		} else {
+			LOG_INF("Erase completed");
+			break;
+		}
+	}
+
+	nrf_close(fd);
+	return 0;
+}
+
+static int do_fota_start(int op, const char *file_uri, int sec_tag,
+			const char *apn)
+{
+	int ret = -EINVAL;
+	struct http_parser_url parser;
+	char schema[8];
+	char path[URI_PATH_MAX];
+	char hostname[URI_HOST_MAX];
+
+	http_parser_url_init(&parser);
+	ret = http_parser_parse_url(file_uri, strlen(file_uri), 0, &parser);
+	if (ret) {
+		LOG_ERR("Parse URL error");
+		return -EINVAL;
+	}
+	memset(schema, 0x00, 8);
+	if (parser.field_set & (1 << UF_SCHEMA)) {
+		strncpy(schema,
+			file_uri + parser.field_data[UF_SCHEMA].off,
+			parser.field_data[UF_SCHEMA].len);
+	} else {
+		LOG_ERR("Parse schema error");
+		return -EINVAL;
+	}
+	memset(path, 0x00, URI_PATH_MAX);
+	if (parser.field_set & (1 << UF_PATH)) {
+		strncpy(path,
+			file_uri + parser.field_data[UF_PATH].off,
+			parser.field_data[UF_PATH].len);
+	} else {
+		LOG_ERR("Parse path error");
+		return -EINVAL;
+	}
+
+	memset(hostname, 0x00, URI_HOST_MAX);
+	strncpy(hostname, file_uri,
+		strlen(file_uri) - parser.field_data[UF_PATH].len);
+
+	/* start HTTP(S) FOTA */
+	if (slm_util_cmd_casecmp(schema, SCHEMA_HTTPS)) {
+		if (sec_tag == INVALID_SEC_TAG) {
+			LOG_ERR("Missing sec_tag");
+			return -EINVAL;
+		}
+		ret = fota_download_start(hostname, path, sec_tag, apn, 0);
+	} else if (slm_util_cmd_casecmp(schema, SCHEMA_HTTP)) {
+		ret = fota_download_start(hostname, path, -1, apn, 0);
+	}
+
+	return ret;
+}
+
+static int do_fota_stop(void)
+{
+	int ret = 0;
```

Comment:
Why is it empty? Shouldn't it come under this `FOTA_FUTURE_FEATURE` define?

Reply:
Exactly, fixed